model/group_syncable.go

```diff
@@ -0,0 +1,165 @@
+// Copyright (c) 2018-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package model
+
+import (
+	"encoding/json"
+	"io"
+	"io/ioutil"
+	"net/http"
+)
+
+type GroupSyncableType int
+
+const (
+	GroupSyncableTypeTeam GroupSyncableType = iota
+	GroupSyncableTypeChannel
+)
+
+func (gst GroupSyncableType) String() string {
+	// Order matters. Keep in sync with iotas.
+	return [...]string{"Team", "Channel"}[gst]
+}
+
+type GroupSyncable struct {
+	GroupId string `json:"group_id"`
+
+	// SyncableId represents the Id of the model that is being synced with the group, for example a ChannelId or
+	// TeamId.
+	SyncableId string `db:"-" json:"-"`
+
+	CanLeave bool              `json:"can_leave"`
+	AutoAdd  bool              `json:"auto_add"`
+	CreateAt int64             `json:"create_at"`
+	DeleteAt int64             `json:"delete_at"`
+	UpdateAt int64             `json:"update_at"`
+	Type     GroupSyncableType `db:"-" json:"-"`
+
+	// Values joined in from the associated team and/or channel
+	ChannelDisplayName string `db:"-" json:"-"`
+	TeamDisplayName    string `db:"-" json:"-"`
+	TeamType           string `db:"-" json:"-"`
+	ChannelType        string `db:"-" json:"-"`
+	TeamID             string `db:"-" json:"-"`
+}
+
+func (syncable *GroupSyncable) IsValid() *AppError {
+	if !IsValidId(syncable.GroupId) {
+		return NewAppError("GroupSyncable.SyncableIsValid", "model.group_syncable.group_id.app_error", nil, "", http.StatusBadRequest)
+	}
+	if !IsValidId(syncable.SyncableId) {
+		return NewAppError("GroupSyncable.SyncableIsValid", "model.group_syncable.syncable_id.app_error", nil, "", http.StatusBadRequest)
+	}
+	if syncable.AutoAdd == false && syncable.CanLeave == false {
+		return NewAppError("GroupSyncable.SyncableIsValid", "model.group_syncable.invalid_state", nil, "", http.StatusBadRequest)
+	}
+	return nil
+}
+
+func (syncable *GroupSyncable) UnmarshalJSON(b []byte) error {
+	var kvp map[string]interface{}
+	err := json.Unmarshal(b, &kvp)
+	if err != nil {
+		return err
+	}
+	for key, value := range kvp {
+		switch key {
+		case "team_id":
+			syncable.SyncableId = value.(string)
+			syncable.Type = GroupSyncableTypeTeam
```

Comment:
I'm not thinking it matters unless I'm missing something?

Reply:
Since the incoming JSON is arbitrary, if I passed a payload of `{"team_id":"a","channel_id":"b"}`, I'd first assign `GroupSyncableTypeTeam` and then `GroupSyncableTypeChannel` which seems like a mistake that the client might want to know about.